mp4parse_capi/src/lib.rs

```diff
@@ -677,15 +694,186 @@ pub unsafe extern fn mp4parse_get_indice_table(parser: *mut mp4parse_parser, tra
+// Convert a 'ctts' compact table to full table by iterator,
+// (sample_with_the_same_offset_count, offset) => (offset), (offset), (offset) ...
+//
+// For example:
+// (2, 10), (4, 9) into (10, 10, 9, 9, 9, 9) by calling next_offset_time().
+struct TimeOffsetIterator<'a> {
+    cur_sample_range: std::ops::Range<u32>,
+    cur_offset: i64,
+    ctts_iter: Option<std::slice::Iter<'a, mp4parse::TimeOffset>>,
+}
+
+impl<'a> Iterator for TimeOffsetIterator<'a> {
+    type Item = i64;
+
+    fn next(&mut self) -> Option<i64> {
+        let has_sample = self.cur_sample_range.next()
+            .or_else(|| {
+                // At end of current TimeOffset, find the next TimeOffset.
+                let iter = match self.ctts_iter {
+                    Some(ref mut v) => v,
+                    _ => return None,
+                };
+                let offset_version;
+                self.cur_sample_range = match iter.next() {
+                    Some(v) => {
+                        offset_version = v.time_offset;
+                        (0 .. v.sample_count)
+                    },
+                    _ => {
+                        offset_version = mp4parse::TimeOffsetVersion::Version0(0);
+                        (0 .. 0)
+                    },
+                };
+
+                self.cur_offset = match offset_version {
+                    mp4parse::TimeOffsetVersion::Version0(i) => i as i64,
+                    mp4parse::TimeOffsetVersion::Version1(i) => i as i64,
+                };
+
+                self.cur_sample_range.next()
+            });
+
+        match has_sample {
+            Some(_) => Some(self.cur_offset),
+            _ => None,
+        }
+    }
+}
+
+impl<'a> TimeOffsetIterator<'a> {
+    fn next_offset_time(&mut self) -> i64 {
+        match self.next() {
+            Some(v) => v as i64,
+            _ => 0,
+        }
+    }
+}
+
+// Convert 'stts' compact table to full table by iterator,
+// (sample_count_with_the_same_time, time) => (time, time, time) ... repeats
+// sample_count_with_the_same_time.
+//
+// For example:
+// (2, 3000), (1, 2999) to (3000, 3000, 2999).
+struct TimeToSampleIteraor<'a> {
+    cur_sample_count: std::ops::Range<u32>,
+    cur_sample_delta: u32,
+    stts_iter: std::slice::Iter<'a, mp4parse::Sample>,
+}
+
+impl<'a> Iterator for TimeToSampleIteraor<'a> {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<u32> {
+        let has_sample = self.cur_sample_count.next()
+            .or_else(|| {
+                self.cur_sample_count = match self.stts_iter.next() {
+                    Some(v) => {
+                        self.cur_sample_delta = v.sample_delta;
+                        (0 .. v.sample_count)
+                    },
+                    _ => (0 .. 0),
+                };
+
+                self.cur_sample_count.next()
+            });
+
+        match has_sample {
+            Some(_) => Some(self.cur_sample_delta),
+            _ => None,
+        }
```

Comment:
FWIW, you can do this in one line with `has_sample.and(Some(self.cur_sample_delta))` but the `match` may be more clear. Up to you.